transport-protocol/src/custom.rs

```diff
@@ -23,35 +35,157 @@ pub enum TransportProtocolError {
     ResponseStatusError(i32),
     #[error(display = "TransportProtocol: TryIntoError: {}.", _0)]
     TryIntoError(#[error(source)] std::num::TryFromIntError),
+    #[error(
+        display = "TransportProtocol: Session {:?}: Buffer is partial, still waiting for chunks",
+        _0
+    )]
+    PartialBuffer(u32),
+    #[error(display = "TransportProtocol: Session {:?}: Length prefix missing", _0)]
+    MissingLengthPrefix(u32),
+    #[error(
+        display = "TransportProtocol: Session {:?}: Incoming buffer not found",
+        _0
+    )]
+    BufferNotFound(u32),
+    #[error(
+        display = "TransportProtocol: Session {:?}: Can't get mutex for incoming buffer",
+        _0
+    )]
+    MutexError(u32),
 }
 type TransportProtocolResult = Result<std::vec::Vec<u8>, TransportProtocolError>;
 
+/// General documentation on the network stack (TODO: move it somewhere else)
+/// Each message is serialized by the sender, prefixed by its length, then gets
+/// encrypted at the TLS layer, then split into 16KB TLS records, sent one by
+/// one to the receiver.
+/// The first chunk therefore contains the protocol buffer's total length.
+/// If the incoming buffer reaches its expected length, the protocol buffer is
+/// considered complete and parsed into a message before flushing the incoming
+/// buffer and returning `Ok(message)`.
+/// Otherwise, the receiver knows it still needs to receive more chunks in order
+/// to parse the buffer, and returns `Ok(None)`.
+/// This technique significantly decreases message parsing time hence latency
+/// when dealing with large messages, e.g. files.
+
+/// Strip the length prefix from the input buffer.
+/// Return the length and the stripped buffer.
+/// This function must be called before deserializing a `protobuf` message
+fn get_length_prefix(buffer: &[u8]) -> (u64, &[u8]) {
+    let mut length_bytes: [u8; LENGTH_PREFIX_SIZE] = [0; LENGTH_PREFIX_SIZE];
+    length_bytes.copy_from_slice(&buffer[..LENGTH_PREFIX_SIZE]);
+    let remaining_buffer = &buffer[LENGTH_PREFIX_SIZE..];
+    (u64::from_be_bytes(length_bytes), remaining_buffer)
+}
+
+/// Return the input buffer prefixed with its length.
+/// This function must be called after serializing a `protobuf` message
+fn set_length_prefix(buffer: &mut Vec<u8>) -> TransportProtocolResult {
+    let length = u64::to_be_bytes(buffer.len() as u64);
+    let mut length_bytes = length.to_vec();
+    length_bytes.append(buffer);
+    Ok(length_bytes)
+}
+
+/// Strip the length prefix from the first chunk received and append subsequent
+/// chunks to the session's incoming buffer until the protocol buffer is
+/// complete and can be deserialized into a message.
+/// This function must be called everytime a new chunk is received.
+/// Take a session id and chunk as input.
+/// Return the complete buffer, or an error indicating that the buffer is
+/// partial, or any other error
+///
+/// TODO: harden this against potential malfeasance.  See the note, below.
```

Comment:
Why is this not already done?

Reply:
The client could saturate the enclave's memory by sending a large file. The amount of enclave memory should be specified in the policy file (#280).
I can't think of any other immediate vulnerabilities here.
Anyway, this PR isn't about hardening the network stack